sdk/core/token_server/alloc.cc

```diff
@@ -0,0 +1,82 @@
+#include <cheri.hh>
+#include <debug.hh>
+#include <token.h>
+
+#include "../allocator/token.h"
+
+using Debug = ConditionalDebug<DEBUG_TOKEN_SERVER, "Token Server">;
+
+using namespace CHERI;
+
+/**
+ * Helper that allocates a sealed object and returns the sealed and
+ * unsealed capabilities to the object.  Requires that the sealing key have
+ * all of the permissions in `permissions`.
+ */
+static std::pair<SObj, void *>
+  __noinline ts_allocate_sealed_unsealed(Token_Allocator heapAlloc,
+                                         Timeout        *timeout,
+                                         SObj            heapCapability,
+                                         SealingKey      key,
+                                         size_t          sz,
+                                         PermissionSet   permissions)
+{
+	if (!permissions.can_derive_from(key.permissions()))
+	{
+		Debug::log(
+		  "Operation requires {}, cannot derive from {}", permissions, key);
+		return {nullptr, nullptr};
+	}
+
+	if (sz > 0xfe8 - ObjHdrSize)
+	{
+		Debug::log("Cannot allocate sealed object of {} bytes, too large", sz);
+		// TODO: Properly handle imprecision.
+		return {nullptr, nullptr};
+	}
+
+	SealedAllocation obj{
+	  static_cast<SObj>(heapAlloc(timeout, heapCapability, sz + ObjHdrSize))};
+	if (obj == nullptr)
+	{
+		Debug::log("Underlying allocation failed for sealed object");
+		return {nullptr, nullptr};
+	}
+
+	obj->type   = key.address();
+	auto sealed = obj;
+	sealed.seal(SEALING_CAP());
+	obj.address() += ObjHdrSize; // Exclude the header.
+	obj.bounds() = obj.length() - ObjHdrSize;
```

Comment:
This chunk probably needs to run with interrupts disabled.  I'm a bit nervous about this decoupling because the security of this code depends on the invariant that no unsealed pointer to the outer object exists anywhere in the system other than in the allocator / token code.  Now that the allocator is untrused, it's trivial to break this.  I can pass a `heapAlloc` function that returns a pointer to something and also stores that pointer in a global, and then I can change the sealing type in the first word.  This makes it trivial for me to fake sealed capabilities.  For example: